sklearn/_engine/tests/test_engines.py

```diff
@@ -136,3 +140,32 @@ def test_get_engine_class():
     next(engine_classes)
     with pytest.raises(ImportError, match=re.escape("sklearn.provider1")):
         next(engine_classes)
+
+
+@pytest.mark.parametrize(
+    "attribute_types,converted", [("sklearn_types", True), ("engine_types", False)]
+)
+def test_attribute_conversion(attribute_types, converted):
+    """Test attribute conversion logic
+
+    The estimator uses Numpy Array API arrays as its native type.
+    """
+
+    class Engine:
+        @staticmethod
+        def convert_to_numpy(name, value):
+            return np.asarray(value)
+
+    class Estimator:
+        _engine_class = Engine
+
+        @convert_attributes
+        def fit(self, X):
+            self.x = np.array_api.asarray(X)
+
+    X = np.array([1, 2, 3])
+    est = Estimator()
+    with config_context(engine_attributes=attribute_types):
+        est.fit(X)
+
+    assert isinstance(est.x, np.ndarray) == converted
```

Comment:
For the sake of naming conventions:

```suggestion
        @convert_attributes
        def fit(self, X):
            self.X_ = np.array_api.asarray(X)

    X = np.array([1, 2, 3])
    est = Estimator()
    with config_context(engine_attributes=attribute_types):
        est.fit(X)

    assert isinstance(est.X_, np.ndarray) == converted
```